core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -392,8 +418,26 @@ private[spark] class ExecutorMonitor(
    * which the `SparkListenerTaskStart` event is posted before the `SparkListenerBlockManagerAdded`
    * event, which is possible because these events are posted in different threads. (see SPARK-4951)
    */
-  private def ensureExecutorIsTracked(id: String): Tracker = {
-    executors.computeIfAbsent(id, _ => new Tracker())
+  private def ensureExecutorIsTracked(id: String, resourceProfileId: Int): Tracker = {
+    val numExecsWithRpId = execResourceProfileCount.computeIfAbsent(resourceProfileId, _ => 0)
+    val execTracker = executors.computeIfAbsent(id, _ => {
+        val newcount = numExecsWithRpId + 1
+        execResourceProfileCount.put(resourceProfileId, newcount)
+        logDebug(s"Executor added with ResourceProfile id: $resourceProfileId " +
+          s"count is now $newcount")
+        new Tracker(resourceProfileId)
+      })
```

Comment:
What is the impact if concurrent updates to execResourceProfileCount (via put) result in inconsistency w.r.t execResourceProfileCount itself and w.r.t executors ? (due to the 'put' happening - across this class).
If we expect execResourceProfileCount for some key to be some value V while updating it to V1, we could use compute() instead and ensure the invariant is satisfied ? Or do we not expect this to happen ?

Reply:
so that shouldn't happen to execResourceProfileCount  because all of the calls to modify it are from the event listener which is single threaded.  There is a call to get that can happen from separate threads but that won't affect the modify case.   We update both executors and execResourceProfileCount at the same time on adding and removing so those should be in lock step.
I think I can change them to compute if it makes it easier to read?

Reply:
If all updates are expected to be single threaded, then change to compute makes no sense (it will simply complicate the code) - I probably missed some note indicating this is the expectation !
Thanks for elaborating Tom.